Api provides admin functionality to add an image

offcast-dev opened the pull request:
PT: https://www.pivotaltracker.com/story/show/178092234
### User Story:
```
As an Api
In order to have the client show specific images for every article
I need to be able to attach an image to an Article
```
___________
### Changes Proposed:
- adds image attachment to article
- sets up base64 decoding
- adds image to article factory
- sets up aws s3 storage 
- seeds articles with unique images
___________
### Packages / Gems Added:
- aws-sdk-s3
___________
### What we Learned:
- how to connect aws storage
- how to attach files
- how to work with base64 encoding
- some mistakes do not appear in automated tests

offcast-dev committed:
adds first test and fixture

DavDan1 committed:
WIP image attachement

1CIM committed:
WIP attaching image
wip add decode service

offcast-dev committed:
attaches image

DavDan1 committed:
configure amazon s3

1CIM committed:
Add aws-sdk-s3 gem
wip setting up serializers for image

offcast-dev committed:
removes def image from serializers

offcast-dev committed:
removes keys before :image in article controller

LucasKnudsen commented on app/controllers/api/articles_controller.rb:
```diff
@@ -42,10 +42,14 @@ def update
   private
 
+  def attach_image(article)
+    params[:article][:image].present? && DecodeService.attach_image(params[:article][:image].first, article.image)
```
We are sending the image as a string (not an array), so the `.first` will fack it up :P

offcast-dev replied:
ok, fixed this

LucasKnudsen commented on spec/requests/api/journalist/journalist_can_create_articles_spec.rb:
```diff
@@ -1,6 +1,9 @@
   let(:journalist) { create(:user, role: 'journalist') }
   let(:auth_headers) { journalist.create_new_auth_token }
+  let(:image) do
+    [ File.read(fixture_path + '/fake-news-fixture.txt') ]
```
Same here with the array, can remove

offcast-dev replied:
and this

offcast-dev replied:
Nah.. doesnt work as a string. Gives this error ActiveSupport::MessageVerifier::InvalidSignature when we create article.
You will have to send an array to us

LucasKnudsen commented on app/serializers/articles_index_serializer.rb:
```diff
@@ -1,5 +1,5 @@
 class ArticlesIndexSerializer < ActiveModel::Serializer
-  attributes :id, :title, :teaser, :date, :category
+  attributes :id, :title, :teaser, :date, :category, :image
```
Are we testing this change? What's the value of image?

offcast-dev replied:
I am not sure about it to be honest. I think image will be an object. In test environment if we try to get to url in serializer (like in PT), all tests fail as it cannot get attribute url of image, but in Rails console url is accessible in article.image.url. I am not really sure what to think about it, probably need to spin up server, create an article, than get one, and look at response. Can do it after dinner

LucasKnudsen commented on app/serializers/articles_show_serializer.rb:
```diff
@@ -1,5 +1,5 @@
 class ArticlesShowSerializer < ActiveModel::Serializer
-  attributes :id, :title, :teaser, :body, :date, :author, :category
+  attributes :id, :title, :teaser, :body, :date, :author, :category, :image
```
And this? Are we sure it's an url?

LucasKnudsen replied:
```rb
def image  
  Rails.env.test? ?
    ActiveStorage::Blob.service.path_for(image.key) 
    : 
    image.service_url(expires_in: 1.hour, disposition: 'inline') 
end
```

LucasKnudsen replied:
Perhaps we need something like this

christerforsgren91 replied:
what goin on here?

offcast-dev replied:
This Amazon remote storage works in development environment but not in test environment. So when we run tests we don't have image url, so the line ```image.service_url(expires_in: 1.hour, disposition: 'inline') ``` errors out. We can check what environment we are in and make serializer return different stuff in different environments. So the tests are green and we get our urls when deployed

LucasKnudsen replied:
Awesum stuff

offcast-dev committed:
removes bug in decode_service

offcast-dev committed:
removes binding.pry

offcast-dev committed:
adds seed with images

offcast-dev committed:
defines image attribute in serializers to return url
serializers are enviroment sensitive, :image contain url in development
refactors image attribute to be a string to work with frontend requests
adds uniqe images to articles when seeding

LucasKnudsen commented on app/controllers/api/articles_controller.rb:
```diff
@@ -43,13 +43,13 @@ def update
 
   def attach_image(article)
-    params[:article][:image].present? && DecodeService.attach_image(params[:article][:image].first, article.image)
+    params[:article][:image].present? && DecodeService.attach_image(params[:article][:image], article.image)
```
So it's working now without the array? :D

offcast-dev replied:
yes, as soon as deployed we will see images on frontend, checked on localhost.

LucasKnudsen commented on app/serializers/articles_show_serializer.rb:
```diff
@@ -11,4 +11,12 @@ def author
     object.created_at.strftime('%F, %H:%M')
   end
+
+  def image
```
Awesome, glad `image` works!

LucasKnudsen commented on db/seeds.rb:
```diff
@@ -39,3 +51,9 @@
                  user_id: user.id)
 end
+
+Article.all.each_with_index do |article, index|
+  file = URI.open(image_url[index])
+  
+  article.image.attach(io: file, filename: 'article_image.jpg', content_type: 'image/jpg')
+end
```
👍

LucasKnudsen reviewed (commented):
My only question would be that is the `image` attribute tested for the other requests?
